yatima-rs/src/typechecker/universe.rs

```diff
@@ -0,0 +1,277 @@
+use std::rc::Rc;
+
+pub type UnivPtr = Rc<Univ>;
+
+/// Indexes are used for environment lookups. They take place of de Bruijn indices or levels.
+pub type Index = usize;
+
+use im::Vector;
+
+/// Universe levels
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Univ {
+  /// Sort 0 aka Prop
+  Zero,
+  /// Sort (n + 1)
+  Succ(UnivPtr),
+  /// Sort (max u v)
+  Max(UnivPtr, UnivPtr),
+  /// Sort (imax u v)
+  IMax(UnivPtr, UnivPtr),
+  /// Sort u
+  Var(Index),
+}
+
+// Reduce, or simplify, the universe levels to a normal form. Notice that universe levels with no free variables always reduce to a number,
+// i.e., a sequence of `Succ`s followed by a `Zero`
+pub fn reduce(univ: &UnivPtr) -> UnivPtr {
+  match &**univ {
+    Univ::Succ(a) => {
+      Rc::new(Univ::Succ(reduce(a)))
+    }
+    Univ::Max(a, b) => {
+      let a = reduce(a);
+      let b = reduce(b);
+      reduce_max(&a, &b)
+    }
+    Univ::IMax(a, b) => {
+      let b = reduce(b);
+      match &*b {
+	// IMax(a, b) will reduce to 0 if b == 0
+	Univ::Zero => b,
```

Comment:
strange indent. Let's add https://github.com/yatima-inc/yatima-lang-alpha/blob/main/rustfmt.toml to standardize formatting